src/main/java/software/amazon/awssdk/crt/s3/S3FinishedResponseContext.java

```diff
@@ -11,48 +11,62 @@ public class S3FinishedResponseContext {
     private final ChecksumAlgorithm checksumAlgorithm;
     private final boolean didValidateChecksum;
 
+    private CrtS3RuntimeException exception;
+
     /*
      * errorCode The CRT error code
      * responseStatus statusCode of the HTTP response
      * errorPayload body of the error response. Can be null if the request completed successfully
      * checksumAlgorithm, the algorithm used to validate the Body, None if not validated
      * didValidateChecksum which is true if the response was validated.
+     * exception CrtS3RuntimeException if any error occurred. May contain the cause of the exception such as a Java exception in a callback.
      */
-    S3FinishedResponseContext(final int errorCode, final int responseStatus, final byte[] errorPayload, final ChecksumAlgorithm checksumAlgorithm, final boolean didValidateChecksum) {
+    S3FinishedResponseContext(final int errorCode, final int responseStatus, final byte[] errorPayload, final ChecksumAlgorithm checksumAlgorithm, final boolean didValidateChecksum, Throwable cause) {
         this.errorCode = errorCode;
         this.responseStatus = responseStatus;
         this.errorPayload = errorPayload;
         this.checksumAlgorithm = checksumAlgorithm;
         this.didValidateChecksum = didValidateChecksum;
+        if (cause != null) {
+            this.exception = new CrtS3RuntimeException(errorCode, responseStatus, errorPayload, cause);
```

Comment:
I am not sure if we want to omit the CRT layer completely from the stack trace, even if the failure was from a Java callback. I do understand the annoyance of nested exceptions, and from a design perspective, we do want to improve error handling in a Java idiomatic way instead of giving access to the CRT error code. See https://github.com/awslabs/aws-crt-java/pull/683#discussion_r1328949219. We should not have exposed the error code directly to begin with, but now we have to maintain backward compatibility.

Debatable: I think it is better for CRT to have a unified exception and the SDK can do `exception.getCause()` to de-nest it where required or pass S3ClientException as a cause to SDKClientException.

Reply:
if Zoe says it's OK to just expose the underlying connection, I'm down with it

Reply:
Discussed it with @TingDaoK as well, and decided to remove CrtS3RuntimeException since the same data was repeated in two places.